javaslang/src/main/java/javaslang/collection/Multimap.java

```diff
@@ -0,0 +1,477 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import javaslang.*;
+import javaslang.control.Option;
+
+import java.util.*;
+import java.util.function.*;
+import java.util.stream.Collector;
+
+/**
+ * An immutable {@code Multimap} interface.
+ *
+ * @param <K> Key type
+ * @param <V> Value type
+ * @author Ruslan Sennov
+ * @since 2.0.0
+ */
+public interface Multimap<K, V> extends Traversable<Tuple2<K, V>>, Function1<K, Traversable<V>>, Kind2<Multimap<?, ?>, K, V> {
+
+    long serialVersionUID = 1L;
+
+    @SuppressWarnings("unchecked")
+    enum ContainerType {
+
+        SET(
+                (Traversable<?> set, Object elem) -> ((Set<Object>) set).add(elem),
+                (Traversable<?> set, Object elem) -> ((Set<Object>) set).remove(elem)
+        ),
+        SORTED_SET(
+                (Traversable<?> set, Object elem) -> ((Set<Object>) set).add(elem),
+                (Traversable<?> set, Object elem) -> ((Set<Object>) set).remove(elem)
+        ),
+        SEQ(
+                (Traversable<?> seq, Object elem) -> ((List<Object>) seq).append(elem),
+                (Traversable<?> seq, Object elem) -> ((List<Object>) seq).remove(elem)
+        );
+
+        final BiFunction<Traversable<?>, Object, Traversable<?>> add;
+        final BiFunction<Traversable<?>, Object, Traversable<?>> remove;
+
+        ContainerType(BiFunction<Traversable<?>, Object, Traversable<?>> add,
+                              BiFunction<Traversable<?>, Object, Traversable<?>> remove) {
+            this.add = add;
+            this.remove = remove;
+        }
```

Comment:
We should make the instance variables `add` and `remove` and the constructor `ContainerType` private.